nemo/core/callbacks.py

```diff
@@ -35,6 +54,463 @@
+        log the loss every 100 steps. Additional scalar tensors can be logged by adding them to the tensors_to_log
+        argument. By default, it always logs the current epoch and the time taken per epoch.
+
+        args:
+            step_freq (int): The frequency of tensorboard logging. Defaults to every 100 steps
+            tensors_to_log (List of str or NmTensor): A list of either tensor names or NmTensors which will be logged
+                every step_freq steps.
+                Defaults to ["loss"] which only prints the loss.
+            wandb_name(str): wandb experiment name.
+                Defaults to None
+            wandb_project(str): wandb project name.
+                Defaults to None
+            args: argparse flags which will be logged as hyperparameters.
+                Defaults to None.
+            log_epoch (bool): Whether to log epoch and epoch training time to tensorboard.
+                Defaults to True.
+        """
+        if not _WANDB_AVAILABLE:
+            logging.error("Could not import wandb. Did you install it (pip install --upgrade wandb)?")
+        self._step_freq = step_freq
+        self._tensors_to_log = tensors_to_log
+        self._name = wandb_name
+        self._project = wandb_project
+        self._args = args
+        self._last_epoch_start = None
+        self._log_epoch = log_epoch
+
+    def on_train_start(self, state):
+        if state["global_rank"] is None or state["global_rank"] == 0:
+            if _WANDB_AVAILABLE and wandb.run is None:
+                wandb.init(name=self._name, project=self._project)
+                if self._args is not None:
+                    wandb.config.update(self._args)
+            elif _WANDB_AVAILABLE and wandb.run is not None:
+                logging.info("Re-using wandb session")
+            else:
+                logging.error("Could not import wandb. Did you install it (pip install --upgrade wandb)?")
+                logging.info("Will not log data to weights and biases.")
+                self._step_freq = -1
+
+    def on_step_end(self, state):
+        # log training metrics
+        if state["global_rank"] is None or state["global_rank"] == 0:
+            if state["step"] % self._step_freq == 0 and self._step_freq > 0:
+                tensors_logged = {t: state["tensors"].get_tensor(t).cpu() for t in self._tensors_to_log}
+                # Always log learning rate
+                tensors_logged['LR'] = state["learning_rate"]
+                self._wandb_log(tensors_logged)
+
+    def on_epoch_start(self, state):
+        if state["global_rank"] is None or state["global_rank"] == 0:
+            self._last_epoch_start = time.time()
+
+    def on_epoch_end(self, state):
+        if state["global_rank"] is None or state["global_rank"] == 0:
+            if self._log_epoch:
+                epoch_time = time.time() - self._last_epoch_start
+                self._wandb_log({"epoch": state["epoch"], "epoch_time": epoch_time})
```

Comment:
In here it makes sense, as we are sending this information to W&B...

Reply:
Or maybe not... seems that I have a problem with understanding what kind of "granularity" of callbacks we expect...

I honestly believe the smaller the better, as then a user can come and cherry-pick only the callbacks that he needs...